# Analysis of the *Installed* Dataset
### Question 1: *Merge the filtered rows (from question 1-3) from the Licence dataset with the installed elevators. How many rows are merged from each dataset ? How many rows does the merged dataset have ?*

In [1]:
import pandas as pd
license = pd.read_csv("../data/raw/licence.csv")
installed = pd.read_json("../data/raw/installed.json")
for df in [installed, license]:
    df.columns = df.columns.str.upper().str.replace(' ', '')

activeLicense = license[license["LICENSESTATUS"] == "ACTIVE"]
merged = pd.merge(installed, activeLicense, on='ELEVATINGDEVICESNUMBER')
print('Amount of rows in the merged dataset: ' + str(len(merged))) 

Amount of rows in the merged dataset: 42665


### Conclusion: *We can now see the amount of rows in the merged dataset. If my understanding is correct we got all rows from the* license *dataset and 42665 rows from the* installed *set.*
- - - 

### Question 2: *Both datasets have a variable about the location of the elevator. Show the observations where both variables are not equal. What do you notice ?*

In [2]:
import numpy as np
diff = merged.loc[merged["LOCATIONOFTHEELEVATINGDEVICE"] != merged["LOCATIONOFDEVICE"]]
print(str(len(diff) * 100 / len(merged)) + '% of rows are different.')

0.3492323918903082% of rows are different.


In [3]:
diff

ELEVATINGDEVICESNUMBER                                     OWNERNAME  \
12                         38  INFRASTRUCTURE ONTARIO AND LANDS CORPORATION   
2896                    13052                  IMH 67 PARKWOODS VILLAGE LTD   
2897                    13053                  IMH 67 PARKWOODS VILLAGE LTD   
3023                    13249                  IMH 70 PARKWOODS VILLAGE LTD   
4292                    15251                                  TRIGLAV APTS   
...                       ...                                           ...   
42325                64718618                          KCAP CENTENNIAL INC.   
42326                64718619                          KCAP CENTENNIAL INC.   
42436                64723611                      SILVER LAKE DEVELOPMENTS   
42437                64723612                      SILVER LAKE DEVELOPMENTS   
42520                64728558                            RYERSON UNIVERSITY   

                                        OWNERADDRESS OWNERACCOUNTNUMBER  \
12               18 KING ST E  TORONTO ON M5C 1C4 CA           redacted   
2896   19 LESMILL RD SUITE 100 TORONTO ON M3B 2T3 CA           redacted   
2897   19 LESMILL RD SUITE 100 TORONTO ON M3B 2T3 CA           redacted   
3023   19 LESMILL RD SUITE 100 TORONTO ON M3B 2T3 CA           redacted   
4292       41 JOHN ST Suite 1 OAKVILLE ON L6K 1G9 CA           redacted   
...                                              ...                ...   
42325    937 PROGRESS AVE  SCARBOROUGH ON M1G 3T8 CA           redacted   
42326    937 PROGRESS AVE  SCARBOROUGH ON M1G 3T8 CA           redacted   
42436        560 WELLINGTON ST  LONDON ON N6A 3R4 CA           redacted   
42437        560 WELLINGTON ST  LONDON ON N6A 3R4 CA           redacted   
42520         350 VICTORIA ST  TORONTO ON M5B 2K3 CA           redacted   

      DEVICECLASS          DEVICETYPE DEVICESTATUS  \
12      Elevators  Passenger Elevator       Active   
2896    Elevators  Passenger Elevator       Active   
2897    Elevators  Passenger Elevator       Active   
3023    Elevators  Passenger Elevator       Active   
4292    Elevators  Passenger Elevator       Active   
...           ...                 ...          ...   
42325   Elevators  Passenger Elevator       Active   
42326   Elevators  Passenger Elevator       Active   
42436   Elevators  Passenger Elevator       Active   
42437   Elevators  Passenger Elevator       Active   
42520   Elevators       LULA Elevator       Active   

                                        LOCATIONOFDEVICE UNDERREVIEW  \
12                                                  None           N   
2896   67 PARKWOODS VILLAGE DR  NORTH YORK M3A 2X4 ON CA           N   
2897   67 PARKWOODS VILLAGE DR  NORTH YORK M3A 2X4 ON CA           N   
3023   70 PARKWOODS VILLAGE DR  NORTH YORK M3A 2X7 ON CA           N   
4292                                                None           N   
...                                                  ...         ...   
42325        937 PROGRESS AVE  SCARBOROUGH M1G 3T8 ON CA           N   
42326        937 PROGRESS AVE  SCARBOROUGH M1G 3T8 ON CA           N   
42436       23 BARREL YARDS BLVD  WATERLOO N2L 0E3 ON CA           N   
42437       23 BARREL YARDS BLVD  WATERLOO N2L 0E3 ON CA           N   
42520                                               None           N   

                           LOCATIONOFTHEELEVATINGDEVICE  \
12                                                  NaN   
2896   67 PARKWOODS VILLAGE DR  DON MILLS M3A 2X4 ON CA   
2897   67 PARKWOODS VILLAGE DR  DON MILLS M3A 2X4 ON CA   
3023   70 PARKWOODS VILLAGE DR  DON MILLS M3A 2X7 ON CA   
4292                                                NaN   
...                                                 ...   
42325           961 PROGRESS AVE  TORONTO M1G 3T5 ON CA   
42326           961 PROGRESS AVE  TORONTO M1G 3T5 ON CA   
42436              110 ERB ST W  WATERLOO N2L 1T5 ON CA   
42437              110 ERB ST W  WATERLOO N2L 1T5 ON 

### Conclusion: *We see here that there are a few rows where we either don't have the address, or that just don't exactly match.*
- - - 

### Question 3: *Can you find elevators where the location of the device is different ? What information from the location can you extract to compare the 2 columns ? What is your suggestion following this analysis ?*

First I tried to see if we could find a correlation for the unmatching rows, but didn't get much more than what we already knew:

In [4]:
diff[["LOCATIONOFTHEELEVATINGDEVICE","LOCATIONOFDEVICE", "OWNERADDRESS", "LICENSEHOLDERADDRESS", "BILLINGADDRESS"]]

LOCATIONOFTHEELEVATINGDEVICE  \
12                                                  NaN   
2896   67 PARKWOODS VILLAGE DR  DON MILLS M3A 2X4 ON CA   
2897   67 PARKWOODS VILLAGE DR  DON MILLS M3A 2X4 ON CA   
3023   70 PARKWOODS VILLAGE DR  DON MILLS M3A 2X7 ON CA   
4292                                                NaN   
...                                                 ...   
42325           961 PROGRESS AVE  TORONTO M1G 3T5 ON CA   
42326           961 PROGRESS AVE  TORONTO M1G 3T5 ON CA   
42436              110 ERB ST W  WATERLOO N2L 1T5 ON CA   
42437              110 ERB ST W  WATERLOO N2L 1T5 ON CA   
42520                                               NaN   

                                        LOCATIONOFDEVICE  \
12                                                  None   
2896   67 PARKWOODS VILLAGE DR  NORTH YORK M3A 2X4 ON CA   
2897   67 PARKWOODS VILLAGE DR  NORTH YORK M3A 2X4 ON CA   
3023   70 PARKWOODS VILLAGE DR  NORTH YORK M3A 2X7 ON CA   
4292                                                None   
...                                                  ...   
42325        937 PROGRESS AVE  SCARBOROUGH M1G 3T8 ON CA   
42326        937 PROGRESS AVE  SCARBOROUGH M1G 3T8 ON CA   
42436       23 BARREL YARDS BLVD  WATERLOO N2L 0E3 ON CA   
42437       23 BARREL YARDS BLVD  WATERLOO N2L 0E3 ON CA   
42520                                               None   

                                        OWNERADDRESS  \
12               18 KING ST E  TORONTO ON M5C 1C4 CA   
2896   19 LESMILL RD SUITE 100 TORONTO ON M3B 2T3 CA   
2897   19 LESMILL RD SUITE 100 TORONTO ON M3B 2T3 CA   
3023   19 LESMILL RD SUITE 100 TORONTO ON M3B 2T3 CA   
4292       41 JOHN ST Suite 1 OAKVILLE ON L6K 1G9 CA   
...                                              ...   
42325    937 PROGRESS AVE  SCARBOROUGH ON M1G 3T8 CA   
42326    937 PROGRESS AVE  SCARBOROUGH ON M1G 3T8 CA   
42436        560 WELLINGTON ST  LONDON ON N6A 3R4 CA   
42437        560 WELLINGTON ST  LONDON ON N6A 3R4 CA   
42520         350 VICTORIA ST  TORONTO ON M5B 2K3 CA   

                                    LICENSEHOLDERADDRESS  \
12                   18 KING ST E  TORONTO ON M5C 1C4 CA   
2896   1 LONGBRIDGE RD 2ND FLOOR THORNHILL ON L4J 1L4 CA   
2897   1 LONGBRIDGE RD 2ND FLOOR THORNHILL ON L4J 1L4 CA   
3023   1 LONGBRIDGE RD 2ND FLOOR THORNHILL ON L4J 1L4 CA   
4292           41 JOHN ST Suite 1 OAKVILLE ON L6K 1G9 CA   
...                                                  ...   
42325  45 ST CLAIR AVE W SUITE 1001 TORONTO ON M4V 1K...   
42326  45 ST CLAIR AVE W SUITE 1001 TORONTO ON M4V 1K...   
42436            560 WELLINGTON ST  LONDON ON N6A 3R4 CA   
42437            560 WELLINGTON ST  LONDON ON N6A 3R4 CA   
42520             350 VICTORIA ST  TORONTO ON M5B 2K3 CA   

                                          BILLINGADDRESS  
12     7755 HURONTARIO ST SUITE 150 BRAMPTON, ON, L6W...  
2896   1 LONGBRIDGE RD 2ND FLOOR THORNHILL, ON, L4J 1...  
2897   1 LONGBRIDGE RD 2ND FLOOR THORNHILL, ON, L4J 1...  
3023   1 LONGBRIDGE RD 2ND FLOOR THORNHILL, ON, L4J 1...  
4292        41 JOHN ST Suite 1 OAKVILLE, ON, L6K 1G9, CA  
...                                                  ...  
42325  45 ST CLAIR AVE W SUITE 1001 TORONTO, ON, M4V ...  
42326  45 ST CLAIR AVE W SUITE 1001 TORONTO, ON, M4V ...  
42436         560 WELLINGTON ST  LONDON, ON, N6A 3R4, CA  
42437         560 WELLINGTON ST  LONDON, ON, N6A 3R4, CA  
42520          350 VICTORIA ST  TORONTO, ON, M5B 2K3, CA  

[149 rows x 5 columns]

As previously mentioned, some addresses just don't match at all, but there are some that are really close. This seems to be a simple formating issue.

A way to not discard those entries would be to compare the street number and exclude only those that do not match.

In [5]:
to_remove = diff.loc[diff["LOCATIONOFTHEELEVATINGDEVICE"].str.extract(r'([^\s]+)').values != diff["LOCATIONOFDEVICE"].str.extract(r'([^\s]+)').values]
to_remove = to_remove.set_index("ELEVATINGDEVICESNUMBER")
merged = merged.set_index("ELEVATINGDEVICESNUMBER")
merged = merged.drop(to_remove.index)

### Conclusion: *I decided to exclude only the rows that either don't have any address, or for which the match is completely off.*
- - - 

### Question 4: *Is there a variable we could use to filter the dataset ? Apply the filter and overwrite the DataFrame*

Similar to what was done with the License dataset, it would make sense to keep only the installed data that has an Active DEVICESTATUS:

In [6]:
merged.groupby("DEVICESTATUS").count()

OWNERNAME  OWNERADDRESS  OWNERACCOUNTNUMBER  \
DEVICESTATUS                                                        
Active                    42276         42276               42276   
Customer Shutdown            17            17                  17   
TSSA Shutdown               295           295                 295   
Undergoing Major Alt          2             2                   2   

                      DEVICECLASS  DEVICETYPE  LOCATIONOFDEVICE  UNDERREVIEW  \
DEVICESTATUS                                                                   
Active                      42276       42276             42276        42276   
Customer Shutdown              17          17                17           17   
TSSA Shutdown                 295         295               295          295   
Undergoing Major Alt            2           2                 2            2   

                      LOCATIONOFTHEELEVATINGDEVICE  \
DEVICESTATUS                                         
Active                                       42276   
Customer Shutdown                               17   
TSSA Shutdown                                  295   
Undergoing Major Alt                             2   

                      ELEVATINGDEVICESLICENSENUMBER  LICENSESTATUS  \
DEVICESTATUS                                                         
Active                                        42276          42276   
Customer Shutdown                                17             17   
TSSA Shutdown                                   295            295   
Undergoing Major Alt                              2              2   

                      LICENSEEXPIRYDATE  LICENSEHOLDER  \
DEVICESTATUS                                             
Active                            42276          42276   
Customer Shutdown                    17             17   
TSSA Shutdown                       295            295   
Undergoing Major Alt                  2              2   

                      LICENSEHOLDERACCOUNTNUMBER  LICENSEHOLDERADDRESS  \
DEVICESTATUS                                                             
Active                                     42276                 42276   
Customer Shutdown                             17                    17   
TSSA Shutdown                                295                   295   
Undergoing Major Alt                           2                     2   

                      BILLINGCUSTOMER  BILLINGADDRESS  BILLINGACCOUNT  
DEVICESTATUS                                                           
Active                          42276           42276           42276  
Customer Shutdown                  17              17              17  
TSSA Shutdown                     295             295             295  
Undergoing Major Alt                2               2               2

In [7]:
merged = merged[merged["DEVICESTATUS"] == 'Active']

### Conclusion: *The dataset has now been updated to only contain Active devices.*
- - - 

### Question 5: *Is there a variable that would need to be cleaned ?*

There sure is! Although the filter we used in the last question would also have been a good fit, there's also the DEVICETYPE columns that could be cleaned up:

In [8]:
merged.groupby("DEVICETYPE").count()

OWNERNAME  OWNERADDRESS  OWNERACCOUNTNUMBER  \
DEVICETYPE                                                          
Freight Elevator           1674          1674                1674   
Freight Elevator-E            8             8                   8   
Freight Elevator-P           11            11                  11   
LULA Elevator              1152          1152                1152   
Material Lift - ATD           1             1                   1   
Observation Elevator        297           297                 297   
Passenger Elevator        39126         39126               39126   
Sidewalk Elevator             2             2                   2   
Special Installation          1             1                   1   
Temporary Elevator            4             4                   4   

                      DEVICECLASS  DEVICESTATUS  LOCATIONOFDEVICE  \
DEVICETYPE                                                          
Freight Elevator             1674          1674              1674   
Freight Elevator-E              8             8                 8   
Freight Elevator-P             11            11                11   
LULA Elevator                1152          1152              1152   
Material Lift - ATD             1             1                 1   
Observation Elevator          297           297               297   
Passenger Elevator          39126         39126             39126   
Sidewalk Elevator               2             2                 2   
Special Installation            1             1                 1   
Temporary Elevator              4             4                 4   

                      UNDERREVIEW  LOCATIONOFTHEELEVATINGDEVICE  \
DEVICETYPE                                                        
Freight Elevator             1674                          1674   
Freight Elevator-E              8                             8   
Freight Elevator-P             11                            11   
LULA Elevator                1152                          1152   
Material Lift - ATD             1                             1   
Observation Elevator          297                           297   
Passenger Elevator          39126                         39126   
Sidewalk Elevator               2                             2   
Special Installation            1                             1   
Temporary Elevator              4                             4   

                      ELEVATINGDEVICESLICENSENUMBER  LICENSESTATUS  \
DEVICETYPE                                                           
Freight Elevator                               1674           1674   
Freight Elevator-E                                8              8   
Freight Elevator-P                               11             11   
LULA Elevator                                  1152           1152   
Material Lift - ATD                               1              1   
Observation Elevator                            297            297   
Passenger Elevator                            39126          39126   
Sidewalk Elevator                                 2              2   
Special Installation                              1              1   
Temporary Elevator                                4              4   

                      LICENSEEXPIRYDATE  LICENSEHOLDER  \
DEVICETYPE                                               
Freight Elevator                   1674           1674   
Freight Elevator-E                    8              8   
Freight Elevator-P                   11             11   
LULA Elevator                      1152           1152   
Material Lift - ATD                   1              1   
Observation Elevator                297            297   
Passenger Elevator                39126          39126   
Sidewalk Elevator                     2              2   
Special Installation                  1              1   
Temporary Elevator                    4              4   

                     

The Passenger Elevator type is clearly the most present in the set. Depending on the context however, it could make sense to keep Freight Elevators and LULA Elevators as well.

In [9]:
merged = merged[merged["DEVICETYPE"].isin(["Passenger Elevator", "Freight Elevator", "LULA Elevator"])]
merged

OWNERNAME  \
ELEVATINGDEVICESNUMBER                                                      
8                       LEGISLATIVE ASSEMBLY OF ONTARIO ATTN:  JOHN ED...   
9                       LEGISLATIVE ASSEMBLY OF ONTARIO ATTN:  JOHN ED...   
10                      LEGISLATIVE ASSEMBLY OF ONTARIO ATTN:  JOHN ED...   
11                      LEGISLATIVE ASSEMBLY OF ONTARIO ATTN:  JOHN ED...   
13                      LEGISLATIVE ASSEMBLY OF ONTARIO ATTN:  JOHN ED...   
...                                                                   ...   
64743755                             ROMAN CATHOLIC EPISCOPAL CORPORATION   
64744591                                                     MILTON REMAX   
64744644                                                         CARTTERA   
64744648                                                         CARTTERA   
64744825                                     OXFORD PROPERTIES CANADA LTD   

                                                             OWNERADDRESS  \
ELEVATINGDEVICESNUMBER                                                      
8                       99 WELLESLEY ST W WHITNEY BLOCK  ROOM 2540 TOR...   
9                       99 WELLESLEY ST W WHITNEY BLOCK  ROOM 2540 TOR...   
10                      99 WELLESLEY ST W WHITNEY BLOCK  ROOM 2540 TOR...   
11                      99 WELLESLEY ST W WHITNEY BLOCK  ROOM 2540 TOR...   
13                      99 WELLESLEY ST W WHITNEY BLOCK  ROOM 2540 TOR...   
...                                                                   ...   
64743755                              71 GOUGH AVE  TORONTO ON M4K 3N9 CA   
64744591                          345 STEELES AVE E  MILTON ON L9T 3G6 CA   
64744644                 20 ADELAIDE ST E SUITE 501 TORONTO ON M5C 2T6 CA   
64744648                 20 ADELAIDE ST E SUITE 501 TORONTO ON M5C 2T6 CA   
64744825                120 ADELAIDE ST W SUITE 702 TORONTO ON M5H 1T1 CA   

                       OWNERACCOUNTNUMBER DEVICECLASS          DEVICETYPE  \
ELEVATINGDEVICESNUMBER                                                      
8                                redacted   Elevators  Passenger Elevator   
9                                redacted   Elevators  Passenger Elevator   
10                               redacted   Elevators  Passenger Elevator   
11                               redacted   Elevators  Passenger Elevator   
13                               redacted   Elevators  Passenger Elevator   
...                                   ...         ...                 ...   
64743755                         redacted   Elevators       LULA Elevator   
64744591                         redacted   Elevators  Passenger Elevator   
64744644                         redacted   Elevators  Passenger Elevator   
64744648                         redacted   Elevators  Passenger Elevator   
64744825                         redacted   Elevators  Passenger Elevator   

                       DEVICESTATUS  \
ELEVATINGDEVICESNUMBER                
8                            Active   
9                            Active   
10                           Active   
11                           Active   
13                           Active   
...                             ...   
64743755                     Active   
64744591                     Active   
64744644                     Active   
64744648                     Active   
64744825                     Active   

                                                       LOCATIONOFDEVICE  \
ELEVATINGDEVICESNUMBER                                                    
8                             111 WELLESLEY ST W  TORONTO M7A 1A2 ON CA   
9                             111 WELLESLEY ST W  TORONTO M7A 1A2 ON CA   
10                            111 WELLESLEY ST W  TORONTO M7A 1A2 ON CA   
11                            111 WELLESLEY ST W  TORONTO M7A 1A2 ON CA   
13                            111 WELLESLEY ST W  TORONTO M7A 1A2 ON CA   
...                      

### Conclusion: *The dataset has now been updated to only contain the most common device type. Again it could be argued that only the Passenger Elevators should be used.*
- - - 

In [10]:
merged.to_csv('../data/processed/License-Installed.csv')